uat/driver/utils.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import subprocess
+
+import tenacity
+from lightkube import Client
+from lightkube.resources.batch_v1 import Job
+from lightkube.resources.core_v1 import Namespace
+
+log = logging.getLogger(__name__)
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=10),
+    stop=tenacity.stop_after_attempt(30),
+    reraise=True,
+)
+def assert_namespace_active(
+    client: Client,
+    namespace: str,
+):
+    """Test that the provided namespace is Active.
+
+    Retries multiple times to allow for the K8s namespace to be created and reach Active status.
+    """
+    # raises a 404 ApiError if the namespace doesn't exist
+    ns = client.get(Namespace, namespace)
+    phase = ns.status.phase
+
+    log.info(f"Waiting for namespace {namespace} to become 'Active': phase == {phase}")
+    assert phase == "Active", f"Waited too long for namespace {namespace}!"
+
+
+def _log_before_sleep(retry_state):
+    """Custom callback to log the number of seconds before the next attempt."""
+    next_attempt = retry_state.attempt_number
+    delay = retry_state.next_action.sleep
+    log.info(f"Retrying in {int(delay)} seconds (attempts: {next_attempt})")
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=64),
+    retry=tenacity.retry_if_not_result(lambda result: result),
+    before_sleep=_log_before_sleep,
+    reraise=True,
+)
+def wait_for_job(
+    client: Client,
+    job_name: str,
+    namespace: str,
+):
+    """Wait for a Kubernetes Job to complete."""
+    # raises a 404 ApiError if the Job doesn't exist
+    job = client.get(Job, name=job_name, namespace=namespace)
+    if job.status.succeeded:
+        # stop retrying, Job succeeded
+        log.info(f"Job {namespace}/{job_name} completed successfully!")
+        return True
+    elif job.status.failed:
+        raise ValueError(f"Job {namespace}/{job_name} failed!")
+    elif not job.status.ready or job.status.active:
+        # continue retrying
+        status = "active" if job.status.active else "not ready"
+        log.info(f"Waiting for Job {namespace}/{job_name} to complete (status == {status})")
+        return False
+    else:
+        raise ValueError(f"Unknown status {job.status} for Job {namespace}/{job_name}!")
```

Comment:
Same comment as [this](https://github.com/canonical/bundle-kubeflow/pull/649/files#r1277743787) - should this be retried, or is this something where we should fail fast?  

Maybe for this one, the current version is good?  Failing fast might be a bit aggressive - perhaps this is a state we dont know about, but that k8s will resolve for us?  I could go either way

Reply:
> Failing fast might be a bit aggressive - perhaps this is a state we dont know about, but that k8s will resolve for us?

I'm more inclined towards this TBH, I don't expect to land here normally, and if I do it means it's most probably a transient state that I did not anticipate.

Reply:
Same note as here: https://github.com/canonical/bundle-kubeflow/pull/649#discussion_r1279246935

With the current implementation, this will fail fast. This is not super important though.

Reply:
lgtm!